src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,370 @@
 **/
 
 #include "UpgradeManager.h"
+#include "libCrypto/Schnorr.h"
 #include "libUtils/Logger.h"
-
+#include <boost/tokenizer.hpp>
+#include <curl/curl.h>
 using namespace std;
 
-SWInfo::SWInfo()
-    : m_major(0)
-    , m_minor(0)
-    , m_fix(0)
-    , m_upgradeDS(0)
-    , m_commit(0)
+#define RELEASE_URL                                                            \
+    "https://api.github.com/repos/Zilliqa/Zilliqa/releases/latest"
+#define USER_AGENT "Zilliqa"
+#define VERSION_FILE_NAME "VERSION"
+#define PUBLIC_KEY_FILE_NAME "pubKeyFile"
+#define PACKAGE_FILE_EXTENSION "deb"
+
+UpgradeManager::UpgradeManager() {}
+
+UpgradeManager::~UpgradeManager() {}
+
+UpgradeManager& UpgradeManager::GetInstance()
 {
+    static UpgradeManager um;
+    return um;
 }
 
-SWInfo::SWInfo(const uint32_t& major, const uint32_t& minor,
-               const uint32_t& fix, const uint64_t& upgradeDS,
-               const uint32_t& commit)
-    : m_major(major)
-    , m_minor(minor)
-    , m_fix(fix)
-    , m_upgradeDS(upgradeDS)
-    , m_commit(commit)
+static size_t WriteString(void* contents, size_t size, size_t nmemb,
+                          void* userp)
 {
+    ((string*)userp)->append((char*)contents, size * nmemb);
+    return size * nmemb;
 }
 
-SWInfo::SWInfo(const SWInfo& src)
-    : m_major(src.m_major)
-    , m_minor(src.m_minor)
-    , m_fix(src.m_fix)
-    , m_upgradeDS(src.m_upgradeDS)
-    , m_commit(src.m_commit)
+static size_t WriteStream(void* ptr, size_t size, size_t nmemb, void* stream)
 {
+    size_t written = fwrite(ptr, size, nmemb, (FILE*)stream);
+    return written;
 }
 
-SWInfo::~SWInfo(){};
-
-/// Implements the Serialize function inherited from Serializable.
-unsigned int SWInfo::Serialize(std::vector<unsigned char>& dst,
-                               unsigned int offset) const
+string UpgradeManager::DownloadFile(const char* fileTail)
 {
-    LOG_MARKER();
+    curl_global_init(CURL_GLOBAL_DEFAULT);
+    CURL* curl = curl_easy_init();
+
+    if (!curl)
+    {
+        curl_easy_cleanup(curl);
+        curl_global_cleanup();
+        return nullptr;
+    }
 
-    unsigned int size_remaining = dst.size() - offset;
+    string curlRes;
+    curl_easy_setopt(curl, CURLOPT_URL, RELEASE_URL);
+    curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteString);
```

Comment:
Seems using a custom callback is not necessary according to https://curl.haxx.se/libcurl/c/libcurl-tutorial.html
> libcurl offers its own default internal callback that will take care of the data if you don't set the callback with CURLOPT_WRITEFUNCTION.

Reply:
Thanks for the information, the WriteStream can be removed. However WriteString would be reserved because here the curl result should be put into string and apply some string tokenization.

Reply:
okay, looks okay to keep `WriteString`